actors/eam/src/lib.rs

```diff
@@ -177,21 +192,56 @@ fn create_actor(
     Ok(Return::from_exec4(ret, new_addr))
 }
 
-fn resolve_caller(rt: &mut impl Runtime) -> Result<EthAddress, ActorError> {
-    let caller_id = rt.message().caller().id().unwrap();
-    Ok(match rt.lookup_delegated_address(caller_id).map(|a| *a.payload()) {
-        Some(Payload::Delegated(addr)) if addr.namespace() == EAM_ACTOR_ID => EthAddress(
+fn resolve_eth_address(rt: &mut impl Runtime, actor_id: ActorID) -> Result<EthAddress, ActorError> {
+    match rt.lookup_delegated_address(actor_id).map(|a| *a.payload()) {
+        Some(Payload::Delegated(addr)) if addr.namespace() == EAM_ACTOR_ID => Ok(EthAddress(
             addr.subaddress()
                 .try_into()
                 .context_code(ExitCode::USR_FORBIDDEN, "caller's eth address isn't valid")?,
-        ),
-        _ => {
-            let mut bytes = [0u8; 20];
-            bytes[0] = 0xff;
-            bytes[12..].copy_from_slice(&caller_id.to_be_bytes());
-            EthAddress(bytes)
+        )),
+        _ => Err(actor_error!(forbidden; "caller doesn't have an eth address")),
+    }
+}
+
+fn resolve_caller_external(rt: &mut impl Runtime) -> Result<(EthAddress, EthAddress), ActorError> {
+    let caller = rt.message().caller();
+    let caller_id = caller.id().unwrap();
+    let caller_code_cid = rt.get_actor_code_cid(&caller_id).expect("failed to lookup caller code");
+    match rt.resolve_builtin_actor_type(&caller_code_cid) {
+        Some(Type::Account) => {
+            let result = rt
+                .send_generalized(
+                    &caller,
+                    PUBKEY_ADDRESS_METHOD,
+                    None,
+                    Zero::zero(),
+                    None,
+                    SendFlags::READ_ONLY,
+                )
+                .context_code(
+                    ExitCode::USR_ASSERTION_FAILED,
+                    "account failed to return its key address",
+                )?;
+
+            if !result.exit_code.is_success() {
+                // TODO: rebase on https://github.com/filecoin-project/builtin-actors/pull/1039
+                return Err(ActorError::unchecked(result.exit_code, format!("failed to call foo")));
+            }
+            let robust_addr: Address = deserialize_block(result.return_data)?;
+            let robust_eth_bytes = hash_20(rt, &robust_addr.to_bytes());
+
+            let mut id_bytes = [0u8; 20];
+            id_bytes[0] = 0xff;
+            id_bytes[12..].copy_from_slice(&caller_id.to_be_bytes());
+
+            Ok((EthAddress(id_bytes), EthAddress(robust_eth_bytes)))
         }
-    })
+        Some(Type::EthAccount) => {
+            let addr = resolve_eth_address(rt, caller_id)?;
+            Ok((addr, addr))
+        }
+        _ => Err(ActorError::forbidden(format!("disallowed caller code cid {}", caller_code_cid))),
```

Comment:
Let's specify the caller type in the error message for better UX/DX.

Reply:
At the cost of a syscall to get the type? I don't think that's the right thing to do, aiding debuggability and UX is fine, but this will cost gas.
The user can always lookup the code CID on his own to find the type.

Reply:
@vyzo we already have the type.

Reply:
ixed.